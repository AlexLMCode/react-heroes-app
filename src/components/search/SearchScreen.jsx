import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';


export const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({
        searchText: q
    });

    const { searchText } = formValues;

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    };

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

    return (
        <div>
            <h3>Search Screen</h3>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSearch}>
                        <input
                            onChange={handleInputChange}
                            type="text"
                            name='searchText'
                            value={searchText}
                            placeholder="Find Your hero"
                            className="form-control"
                        />
                        <button type="submit" className="btn mt-1 btn-primary">Search</button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results:</h4>

                    {
                        (q === '') &&

                        <div className="alert alert-primary" role="alert">
                            Search a Hero
                            </div>

                    }

                    {
                        (q !== '' && heroesFiltered.length === 0) &&

                        <div className="alert alert-danger" role="alert">
                            There is no a hero with {q}
                        </div>
                    }


                    {
                        heroesFiltered.map(hero => (
                            <HeroCard {...hero} key={hero.id} />
                        ))
                    }

                </div>
            </div>
        </div>
    );
};
