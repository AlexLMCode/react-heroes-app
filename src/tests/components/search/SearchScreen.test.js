import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Tests in <SearchScreen/>', () => {

    const historyMock = {
        push: jest.fn()
    };

    test('should render with the default values', () => {

        const wrapper = mount(
            <MemoryRouter>
                <Route to="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-primary').text()).toBe('Search a Hero');

    });

    test('should show Batman and the input with the value of the query string', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Batman']}>
                <Route to="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('Batman');
        expect(wrapper).toMatchSnapshot();

    });

    test('should show an error if there is no hero with that query', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Batmankjefe']}>
                <Route to="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text()).toBe("There is no a hero with Batmankjefe");

    });

    test('should call the push of history', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Batmankjefe']}>
                <Route to="/search" component={() => <SearchScreen history={historyMock} />} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`)

    });

});
