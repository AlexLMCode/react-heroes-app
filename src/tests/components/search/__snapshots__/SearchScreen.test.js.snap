// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen/> should render with the default values 1`] = `
<div>
  <h3>
    Search Screen
  </h3>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find Your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results:
      </h4>
      <div
        className="alert alert-primary"
        role="alert"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests in <SearchScreen/> should show Batman and the input with the value of the query string 1`] = `
<div>
  <h3>
    Search Screen
  </h3>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find Your hero"
          type="text"
          value="Batman"
        />
        <button
          className="btn mt-1 btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results:
      </h4>
      <div
        className="card mb-3 animate__animated animate__fadeInUpBig"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <img
          alt="Batman"
          className="card-img-top"
          src="./assets/heroes/dc-batman.jpg"
        />
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            Batman
          </h5>
          <h6
            className="card-subtitle mb-2 text-muted"
          >
            DC Comics
          </h6>
          <p
            className="card-text"
          >
            <small>
              Detective Comics #27
            </small>
          </p>
          <a
            className="btn btn-primary"
            href="/heroe/dc-batman"
            onClick={[Function]}
          >
            See Hero
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
